### PR TITLE
CloudWriter: local=, remote=, keep=.

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:

```
with MDSWriter(columns=..., local=..., remote=..., keep=..., ...) as out:
    for sample in samples:
        out.write(samples)
```

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [ ] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/format/mds/writer.py [hunk 1]
```diff
@@ -19,27 +19,42 @@ class MDSWriter(JointWriter):
...
     Args:
-        dirname (str): Local dataset directory.
         columns (Dict[str, str]): Sample columns.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
+            before uploading. One or both of ``local`` and ``remote`` must be provided. Defaults to
+            ``None``.
+        remote: (str, optional): Optional remote output dataset directory. If not provided, no
+            uploading will be done. Defaults to ``None``.
+        keep (bool): If the dataset is uploaded, whether to keep the local dataset directory  or
```
Comment (karan6181) [inline streaming/base/format/mds/writer.py:29]: maybe `keep_local` as a better name ?

File: streaming/base/format/mds/writer.py [hunk 2]
```diff
@@ -19,27 +19,42 @@ class MDSWriter(JointWriter):
     """Writes a streaming MDS dataset.
 
     Args:
-        dirname (str): Local dataset directory.
         columns (Dict[str, str]): Sample columns.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
```
Comment (karan6181) [inline streaming/base/format/mds/writer.py:23]: Just thinking out loud, should we also have an another optional parameter called src_dir where it contains the raw source files which can be image, text, parquet, etc. ? Then Writer converts the files into a .mds shard files in a `local` directory?

File: streaming/base/format/base/writer.py [hunk 3]
```diff
@@ -197,21 +249,26 @@ def write(self, sample: Dict[str, Any]) -> None:
 
     def _write_index(self) -> None:
         """Write the index, having written all the shards."""
-        assert not self.new_samples
-        filename = os.path.join(self.dirname, get_index_basename())
+        if self.new_samples:
+            raise RuntimeError('Internal error.')
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:253]: Can we please add more descriptive error message ?

File: streaming/vision/convert/coco.py [hunk 4]
```diff
@@ -235,7 +235,11 @@ def main(args: Namespace) -> None:
         else:
             dataset = each(dataset, shuffle)
 
-        with MDSWriter(split_out_dir, fields, args.compression, hashes, args.size_limit) as out:
+        with MDSWriter(local=split_out_dir,
```
Comment (karan6181) [inline streaming/vision/convert/coco.py:238]: nit: `split_dir` in sync with others ?

File: streaming/vision/convert/ade20k.py [hunk 5]
```diff
@@ -153,14 +153,17 @@ def main(args: Namespace) -> None:
...
-        split_images_out_dir = os.path.join(args.out_root, split)
+        out_dir = os.path.join(args.out_root, split)
         hashes = get_list_arg(args.hashes)
 
         if args.progbar:
             samples = tqdm(samples, leave=args.leave)
 
-        with MDSWriter(split_images_out_dir, fields, args.compression, hashes,
-                       args.size_limit) as out:
+        with MDSWriter(local=out_dir,
```
Comment (karan6181) [inline streaming/vision/convert/ade20k.py:162]: nit: `split_dir` in sync with others ?

File: streaming/text/convert/pile.py [hunk 6]
```diff
@@ -145,7 +145,11 @@ def file_to_dir(args: Tuple[str, str, str, List[str], int]) -> Dict[str, int]:
     }
 
     counts = Counter()
-    with MDSWriter(out_dir, columns, compression, hashes, size_limit) as out:
+    with MDSWriter(local=out_dir,
```
Comment (karan6181) [inline streaming/text/convert/pile.py:148]: nit: `split_dir` in sync with others ?

File: streaming/multimodal/convert/webvid.py [hunk 7]
```diff
@@ -148,7 +148,7 @@ def main(args: Namespace) -> None:
     todos = each_todo(getattr(args, 'in'))
     if args.limit:
         todos = head(todos, args.limit)
-    with MDSWriter(args.out, columns) as out:
+    with MDSWriter(local=args.out, columns=columns) as out:
```
Comment (karan6181) [inline streaming/multimodal/convert/webvid.py:151]: nit: `split_dir` in sync with others ?

File: streaming/multimodal/convert/laion/laion400m/convert_and_upload.py [hunk 8]
```diff
@@ -145,7 +145,11 @@ def convert(parquet_filename: str, mds_dirname: str, hashes: List[str]) -> None:
     compression = None  # Don't compress because the vast majority of the data is JPEG.
     size_limit = None  # Put everything in one shard (1:1 mapping of parquet to MDS shards).
 
-    with MDSWriter(mds_dirname, columns, compression, hashes, size_limit) as out:
+    with MDSWriter(local=mds_dirname,
```
Comment (karan6181) [inline streaming/multimodal/convert/laion/laion400m/convert_and_upload.py:148]: nit: `split_dir` in sync with others ?

File: streaming/base/format/xsv/writer.py [hunk 9]
```diff
@@ -154,29 +185,46 @@ class TSVWriter(XSVWriter):
     r"""Writes a streaming TSV dataset.
 
     Args:
-        dirname (str): Local dataset directory.
         columns (Dict[str, str]): Sample columns.
+        newline (str): Newline character inserted between samples. Defaults to ``\\n``.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/xsv/writer.py:191]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/xsv/writer.py [hunk 10]
```diff
@@ -114,29 +128,46 @@ class CSVWriter(XSVWriter):
     r"""Writes a streaming CSV dataset.
 
     Args:
-        dirname (str): Local dataset directory.
         columns (Dict[str, str]): Sample columns.
+        newline (str): Newline character inserted between samples. Defaults to ``\\n``.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/xsv/writer.py:134]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/xsv/writer.py [hunk 11]
```diff
@@ -18,30 +18,44 @@ class XSVWriter(SplitWriter):
     r"""Writes a streaming XSV dataset.
 
     Args:
-        dirname (str): Local dataset directory.
         columns (Dict[str, str]): Sample columns.
         separator (str): String used to separate columns.
+        newline (str): Newline character inserted between samples. Defaults to ``\\n``.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/xsv/writer.py:25]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/mds/writer.py [hunk 12]
```diff
@@ -19,27 +19,42 @@ class MDSWriter(JointWriter):
...
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/mds/writer.py:24]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/json/writer.py [hunk 13]
```diff
@@ -18,28 +18,42 @@ class JSONWriter(SplitWriter):
     r"""Writes a streaming JSON dataset.
 
     Args:
-        dirname (str): Local dataset directory.
         columns (Dict[str, str]): Sample columns.
+        newline (str): Newline character inserted between samples. Defaults to ``\\n``.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/json/writer.py:24]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/base/writer.py [hunk 14]
```diff
@@ -282,14 +355,23 @@ def flush_shard(self) -> None:
...
 
 class SplitWriter(Writer):
     """Writes a streaming dataset with split shards.
 
     Split shards refer to raw data (csv, json, etc.) paired with an index into it.
 
     Args:
-        dirname (str): Local dataset directory.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:368]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/base/writer.py [hunk 15]
```diff
@@ -237,7 +294,14 @@ class JointWriter(Writer):
     """Writes a streaming dataset with joint shards.
 
     Args:
-        dirname (str): Local dataset directory.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:298]: nit: extra space between `temp` and `directory`.

File: streaming/base/format/base/writer.py [hunk 16]
```diff
@@ -38,27 +60,43 @@ class Writer(ABC):
...
+                 *,
+                 local: Optional[str] = None,
+                 remote: Optional[str] = None,
+                 keep: bool = False,
                  compression: Optional[str] = None,
                  hashes: Optional[List[str]] = None,
                  size_limit: Optional[int] = 1 << 26,
                  extra_bytes_per_shard: int = 0,
                  extra_bytes_per_sample: int = 0) -> None:
+        if local:
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:72]: can we please write specifically `if local is None` ?

File: streaming/base/format/base/writer.py [hunk 17]
```diff
@@ -38,27 +60,43 @@ class Writer(ABC):
...
+            pass
+        elif remote:
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:74]: can we please write specifically `if remote is None` ?

File: streaming/base/format/base/writer.py [hunk 18]
```diff
@@ -18,11 +19,32 @@
...
+    shutil.copy(local, remote)
+
+
 class Writer(ABC):
     """Writes a streaming dataset.
 
     Args:
-        dirname (str): Local dataset directory.
+        local: (str, optional): Optional local output dataset directory. If not provided, a random
+           temp  directory will be used. If ``remote`` is provided, this is where shards are cached
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:41]: extra space

**Review by karan6181 — COMMENTED** [review]
Great changes!

File: streaming/base/format/mds/writer.py [hunk 19]
```diff
@@ -19,27 +19,42 @@ class MDSWriter(JointWriter):
...
+            before uploading. One or both of ``local`` and ``remote`` must be provided. Defaults to
+            ``None``.
+        remote: (str, optional): Optional remote output dataset directory. If not provided, no
```
[the remark this answers is not in the document]
Comment (knighton) [reply streaming/base/format/mds/writer.py:27]: yes

File: streaming/base/format/mds/writer.py (hunk 2 above)
Comment (knighton) [reply streaming/base/format/mds/writer.py:23]: we can't make assumptions about where samples are coming from

File: streaming/text/convert/pile.py (hunk 6 above)
Comment (knighton) [reply streaming/text/convert/pile.py:148]: here we have `in_file` and `out_dir`

File: streaming/vision/convert/coco.py (hunk 4 above)
Comment (knighton) [reply streaming/vision/convert/coco.py:238]: other variables are named `split_..._in_dir`

File: streaming/vision/convert/ade20k.py (hunk 5 above)
Comment (knighton) [reply streaming/vision/convert/ade20k.py:162]: here we have `in_root`, `out_root`, `split`, so `split_dir` would be confusing

File: streaming/multimodal/convert/webvid.py (hunk 7 above)
Comment (knighton) [reply streaming/multimodal/convert/webvid.py:151]: this is not a split dir

File: streaming/multimodal/convert/laion/laion400m/convert_and_upload.py (hunk 8 above)
Comment (knighton) [reply streaming/multimodal/convert/laion/laion400m/convert_and_upload.py:148]: here we have `parquet_filename` and `mds_dirname`

File: streaming/base/format/base/writer.py (hunk 16 above)
Comment (knighton) [reply streaming/base/format/base/writer.py:72]: empty string should be falsy

File: streaming/base/format/base/writer.py (hunk 17 above)
Comment (knighton) [reply streaming/base/format/base/writer.py:74]: empty string should be falsy

**Review by karan6181 — APPROVED** [review]

**Pull request merged**